src/Microsoft.AspNetCore.Razor.Language/Syntax/MarkupElementRewriter.cs

```diff
@@ -0,0 +1,352 @@
+                            if (TryRecoverStartTag(node, tagName, tagBlock, currentIndex, out var rewritten, out var startTagIndex))
+                            {
+                                node = rewritten;
+
+                                // Set the index back to the start tag's position.
+                                currentIndex = startTagIndex;
+                            }
+                            else
+                            {
+                                // Could not recover. The end tag doesn't have a corresponding start tag. Wrap it in a block and move on.
+                                node = RewriteNodeCore(node, startTag: null, tagChildren: Enumerable.Empty<RazorSyntaxNode>(), endTag: tagBlock);
+                            }
+                        }
+
+                        // Since we rewrote 'node', it's children are different. Update our collection.
+                        children = node.ChildNodes();
+                    }
+                    else
+                    {
+                        // This is a start tag. Keep track of it.
+                        _startTagTracker.Push(new TagBlockTracker(tagBlock, node, currentIndex));
+                    }
+
+                    currentIndex++;
+                }
+
+                while (_startTagTracker.Count > 0)
+                {
+                    // We reached the end of the list and still have unmatched start tags
+                    var startTagTracker = _startTagTracker.Pop();
+                    var startTagIndex = startTagTracker.Index;
+                    var startTag = (MarkupTagBlockSyntax)children[startTagIndex];
+                    var tagChildren = children.Skip(startTagIndex + 1).Cast<RazorSyntaxNode>();
+
+                    node = RewriteNodeCore(node, startTag, tagChildren, endTag: null);
+
+                    // Since we rewrote 'node', it's children are different. Update our collection.
+                    children = node.ChildNodes();
+                }
+
+                return node;
+            }
+
+            private SyntaxNode RewriteNodeCore(
+                SyntaxNode parent,
+                MarkupTagBlockSyntax startTag,
+                IEnumerable<RazorSyntaxNode> tagChildren,
+                MarkupTagBlockSyntax endTag)
+            {
+                // Build a list of the original nodes that we want to replace with the rewritten element.
+                var originalNodes = new List<SyntaxNode>();
+                if (startTag != null)
+                {
+                    originalNodes.Add(startTag);
+                }
+                originalNodes.AddRange(tagChildren);
+                if (endTag != null)
+                {
+                    originalNodes.Add(endTag);
+                }
+
+                if (originalNodes.Count == 0)
+                {
+                    // Nothing to replace
+                    return parent;
+                }
```

Comment:
Does this preserve order correctly? What about a case like:
```
<div>Hi</div>
@Something!
<div>bye</div>
```

Reply:
I see, you did the below.

Reply:
I think this would make a little more sense if you mutated `originalNodes` in place and preserved ordering while doing so.